Structure changes

one-some opened the pull request:
This PR:
- Obliterates `utils.HAS_ACCELERATE` to make maintaining code easier. Accelerate has been a requirement for a while.
- Splits up `load_model` and other `aiserver.py` inference code into a million bajillion pieces: 
    - Models are now represented by a `InferenceModel`-based class with API-user-facing `load()`, `core_generate()` and `raw_generate()` functions. Each model should be located in its own `modeling/inference_models/SUPER_DUPER_MODEL.py` file.
    - The previously transformers-only StoppingCriteria stoppers and token hooks have been replaced with generic `Stoppers` and `PostTokenHooks` methods. These can be implemented on any model type with relative ease and a reference implementation can be found in `modeling/inference_models/hf_torch.py`.
    - Samplers are now much more universal and are no longer tied to Transformers (as far as I can see). All sampler code is now located in `modeling/warpers.py` and each sampler is structured into a singleton class with `torch`, `jax_dynamic`, `jax_static`, and `value_is_valid` methods. This should make creating samplers easier, but more importantly, make implementing them into unsupported models easier.
- Zaps the old, hacky, and crusty RWKV implementation into oblivion. Nobody (to my knowledge) was using it due to the sheer impracticality and its concealment behind the Experimental toggle. A new implementation is present and more functional than the last one, but still not user-ready, and thus hidden behind the Experimental toggle.
- [Moves `aetherroom.club` import code out of aiserver.py](https://github.com/one-some/KoboldAI-united/pull/1) (by @whjms)
- Adds unit tests for model loading and inference (`modeling/test_generation.py`)
- Makes modellist (in `aiserver.py`) oriented around objects rather than lists. This makes working with and expanding the model menu a lot easier
- [Adds API endpoint for custom stop sequences](https://github.com/one-some/KoboldAI-united/pull/5) (by @LostRuins)

This will need a lot of testing. I have been testing extensively (and will continue to), but the bugs have started to look normal to me at this point :^)

LightSaveUs committed:
RTE (Custom)

LightSaveUs committed:
RTE (13B)

LightSaveUs committed:
RTE (6B)

Commit:
Model: Successful load implementation

The goal of this series of commits is to have an implementation-agnostic
interface for models, thus being less reliant on HF Transformers for model
support. A model object will have a method for generation, a list of callbacks
to be run on every token generation, a list of samplers that will modify
probabilities, etc. Basically anything HF can do should be easily
implementable with the new interface :^)

Currently I've tested the loading of pre-downloaded models with
breakmodel between GPUs and that works, though essentially no testing
has been done in the larger scheme of things. Currently this is about
the only supported configuration, and generation isn't very functional.

Commit:
Model: Port rest of models over

Generation's still broke but it's a start

Commit:
Model: Work on stoppers and stuff

Commit:
Model Fix bugs and introduce hack for visualization

Hopefully I remove that attrocity before the PR

Commit:
Model: Fix crash on full GPU load

Commit:
Model: More TPU stuff

Commit:
Model: Tokenizer fix

Commit:
Accelerate: Remove HAS_ACCELERATE

Accelerate has been a dependency for a while, and as such we probably
shouldn't be lugging around code that assumes it isn't present.

Commit:
Model: Fix ReadOnly

whoops

Commit:
Merge branch 'united' of https://github.com/henk717/KoboldAI into model-structure-and-maybe-rwkv

Commit:
Model: Monkey release detection fix

Commit:
Model: Add rep pen for OAI API

Commit:
Model: Fixup horde bug

Commit:
Get rid of yet another instance of this identical class

Commit:
Model: Formatting fixes

Commit:
Model: Samplers pt. 1

Commit:
Samplers: Part 2

Commit:
Model: Bugfixes/fix tokenizer hack

Commit:
colab debug

Commit:
Model: TPU should be ready for testing

Commit:
Model: Don't do breakmodel on colab

Commit:
Model: More tpu fixes

Commit:
Model: TPU Fixes

Commit:
Model: Documentation part 1

henk717 committed:
Nerybus 13B - TPU colab

Commit:
Model: And another refactor

Commit:
RWKV: Remove old gnarly (in a bad way) RWKV4 support

Better interface for it coming soon

Commit:
Model: Fix TPU

henk717 committed:
Remote no longer unblocks the port by default

henk717 committed:
Enable BMS API again under a new name

henk717 committed:
Remove old worker API

Commit:
Model: Small cleanup

Commit:
Model: Add singleline_stopper and fix stopper code

singleline_stopper adapted from MasterAibo in 0ba7ac9

Commit:
oops

Commit:
Model: Add debug code for detecting faulty samplers

Commit:
Model: Fix API support

Commit:
Attempts at dynamic wi fixes

Commit:
Model: Refuse to serve certain models over the API

Commit:
Merge branch 'united' of https://github.com/henk717/KoboldAI into model-structure-and-maybe-rwkv

Commit:
Model: Rename "Colab" to "Basic API"

one-some commented:
This should be ready for review

one-some commented:
...or not. found a particularly nasty bug with sampling or determinism or something. whatever the case its busted. probably skipping over some hf score manipulation thing by overriding it with our samplers

Commit:
Model: Respect sampler bounds in torch

A rather embarassing way to spend an hour debugging after I told myself
"I'd better remember to add this important thing to the torch side".

Samplers were being applied when in their "off values" causing
boring mathmatical operations to take place (ie anything x 0 is always
0)

one-some commented:
ok now its ready for real (no fake)

Commit:
ignore .venv

Commit:
move aetherroom import to separate module

Commit:
fix worldinfo prop name

Cohee1207 committed:
Fix setting softprompt via V1 API

Commit:
fix 404 handler

Commit:
remove black dep

Commit:
Don't recurse over .git for pytest

whjms commented on modeling/inference_model.py:
```diff
@@ -0,0 +1,601 @@
+
+# I don't really like this way of pointing to the current model but I can't
+# find a way around it in some areas.
+current_model = None
```
would having `HFTorchInferenceModel` just access the tokenizer with `self.tokenizer` work?

one-some replied:
looks like it! i think this was a byproduct of originally not having the patch functions be methods of `HFTorchInferenceModel`, but now they are. good catch!

Commit:
Model: Ditch awful current_model hack

thanks to whjms for spotting that this could be zapped

Commit:
Model: Reformat and clean up

Commit:
Cleanup

Commit:
Model: Respect model lazyload over kaivars

kaivars dictates model config unless its from outside aiserver or
whatever.

Commit:
Model: Fix assorted bugs

and ignore warnings in pytest

Commit:
Model: Add basic tests

We now do some basic tests for:
- hf torch loading (normal, lazy, lowmem)
- hf torch generation (shape batches, shape tokencount, faulty
  determinism)

Currently full determinism is failing; yahoo, the tests work!

All of the tests initally failed (note the test environment functions
different from the aiserver environment due to aiserver doing a lot of
initalizing stuff, working on phasing that out) but now only one fails.
Very useful for finding bugs!

Commit:
Model: Fix eos/bos padding issue

Weird config None assignments

Commit:
Remove ancient debug

Commit:
Modeling: Add seed parameter to raw_generate

Yahooo, decoupling from koboldai_vars. This makes the generation test
pass in `test_generation.py`, and makes full determinism outside of
core_generate work.

Commit:
Model: WIP horde and API tests

Commit:
Make modellist easier to work with

Commit:
Model: Add GenericTokenizer

Because Hugging Face doesnt have a consistant API across their own
libraries

Commit:
Model: Add basic RWKV implementation

Commit:
RWKV: Fix typo

Commit:
RWKV: Move import right before usage

So we don't needlessly compile the cuda kernel

Commit:
RWKV: Fix yet another typo

Commit:
TPUMTJ: Fix loading bar

I don't know why it works but I know it works

Commit:
Model: Lazyload backends

Commit:
Undo debug

Commit:
Modeling: Maybe fix samplers

Commit:
Model: More Jax import fixes and formatting

Commit:
Merge branch 'united' of https://github.com/henk717/KoboldAI into model-structure-and-maybe-rwkv

Commit:
Modeling: Fix logits processors (probs, biasing, lua)

Commit:
Add small debug for TPU

Commit:
GenericTokenizer: Fall back to defined tokenizer

Shouldn't be relied on for model-agnostic code, but for loading
processes where you know the tokenizer class used it should be okie
dokie

Commit:
Lazyload: Safetensors

Commit:
Model fix

Commit:
Merge branch 'united' of https://github.com/henk717/KoboldAI into model-structure-and-maybe-rwkv

Commit:
Breakmodel: Fix typo

Commit:
Small fixes

Typos galore!

Commit:
Merge branch 'united' of https://github.com/henk717/KoboldAI into model-structure-and-maybe-rwkv

Commit:
Fix for custom gpt2

Commit:
Fix for tokenizer stuff on pythia

Commit:
possible fix for cache dl thing

Commit:
Merge branch 'united' of https://github.com/henk717/KoboldAI into model-structure-and-maybe-rwkv

LostRuins committed:
increase API version (+1 squashed commits)

Squashed commits:

[c168c08] Added stop sequences functionality for API calls

LostRuins committed:
Added stop sequences functionality for API calls

LostRuins committed:
increase API version

LostRuins committed:
Merge branch 'concedo_api' of https://github.com/LostRuins/KoboldAI into concedo_api

Commit:
Remove line that sets disk_layers to None always

whoops

Commit:
Remove legacy no accelerate fallback code

Was causing issues with disk cache the old code had a
`and not utils.HAS_ACCELERATE` preceding it (a variable which no longer
exists), and since disk cache is accelerate only, there was no disk
handling code in here. Anyway its bad so blast it

one-some committed:
Merge pull request #5 from LostRuins/concedo_api

Added stop sequences functionality for API calls

Commit:
Utils: Support safetensors aria2 download

Commit:
Merge branch 'model-structure-and-maybe-rwkv' of https://github.com/one-some/KoboldAI into model-structure-and-maybe-rwkv

Commit:
Fix legacy model loading

Commit:
Fix api

Commit:
Fix double slashing

Commit:
Workaround for socketio context errors for loading

Commit:
Remove debug

one-some committed:
Merge branch 'united' of https://github.com/henk717/KoboldAI into model-structure-and-maybe-rwkv

one-some committed:
Make RWKV experimental

one-some committed:
Change fallback notifications to warnings

one-some committed:
Use safetensors only when available

one-some committed:
Lazyload: Whoops

one-some committed:
Undo pretty code because I haven't cracked the jax enigma yet

one-some committed:
Zap debug

one-some committed:
Better fallback

one-some committed:
Delete legacy gpt2 custom loader

one-some committed:
Fix custmodpth stuff for hf loading

one-some committed:
Add more info to custom model error

one-some committed:
Fix for multiple paths

one-some committed:
More loading fixes

Commit:
Cut out things until it works

Commit:
Debug

Commit:
Potential fix

Commit:
Debuuuug

Commit:
Experiment

Commit:
Actually fix decoding with soft prompts

it really wants a tensor

Commit:
One more time

Commit:
Two more time

Commit:
Hello Colab

Commit:
More colab

Commit:
Potential fix for tpu dynamic

Commit:
Implement softprompt hack

Commit:
Dynamic Fix

Commit:
Remove some debug

Commit:
Merge branch 'united' of https://github.com/henk717/KoboldAI into model-structure-and-maybe-rwkv

Commit:
HFInferenceModel: Make badwordsids not unique to torch

Commit:
Download all shards correctly on aria2 and raise on bad load key

Commit:
Support safetensors in get_sharded_checkpoint_num_tensors

Commit:
Fix tokenizer fallback for llama
